app/src/main/java/com/rizadwi/snapsift/viewmodel/ArticleViewModel.kt

```diff
@@ -0,0 +1,110 @@
+import kotlinx.coroutines.flow.collectLatest
+import kotlinx.coroutines.flow.debounce
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+@HiltViewModel
+class ArticleViewModel @Inject constructor(private val getHeadlineArticlesUserCase: GetHeadlineArticlesUserCase) :
+    ViewModel() {
+    private val _flow = MutableSharedFlow<String>()
+    private val _articleLiveData = MutableLiveData<UIState<List<Article>>>()
+    private val _freshArticleLiveData = MutableLiveData<UIState<List<Article>>>()
+
+    private var cachedArticles = mutableListOf<Article>()
+    private var currentPage = FIRST_PAGE
+    private var query = ""
+    private var sources = ""
+
+    val articleLiveData: LiveData<UIState<List<Article>>> = _articleLiveData
+    val freshArticleLiveData: LiveData<UIState<List<Article>>> = _freshArticleLiveData
+
+    init {
+        viewModelScope.launch {
+            _flow.debounce(DELAY).collectLatest(::resetThenLoadFreshArticle)
+        }
+    }
+
+    fun loadFromCacheIfPossible() = viewModelScope.launch {
+        if (cachedArticles.isEmpty()) {
+            loadFreshHeadlineArticles()
+        } else {
+            _freshArticleLiveData.postSuccess(cachedArticles)
+        }
+    }
+
+    private fun loadFreshHeadlineArticles() = viewModelScope.launch {
+        loadHeadlineArticles(_freshArticleLiveData)
+    }
+
+    fun loadMoreHeadlineArticles() = viewModelScope.launch {
+        loadHeadlineArticles(_articleLiveData)
+    }
+
+    fun changeQueryThenLoadFreshHeadlineArticles(query: String) = viewModelScope.launch {
+        _flow.emit(query)
+    }
+
+    fun setSources(sources: String) {
+        this.sources = sources
+    }
+
+    private fun resetThenLoadFreshArticle(keyword: String) {
+        // Ignore if the keyword still the same as previous query
+        if (query == keyword) {
+            return
+        }
+
+        query = keyword
+        currentPage = FIRST_PAGE
+        cachedArticles = mutableListOf()
+
+        loadFreshHeadlineArticles()
+    }
+
+    private fun loadHeadlineArticles(liveData: MutableLiveData<UIState<List<Article>>>) =
+        viewModelScope.launch {
+            liveData.postLoading()
+
+            if (sources.isBlank()) {
+                liveData.postError(Error("Sources should be specified"))
+                return@launch
+            }
+
+
+            val result = if (query.isBlank()) {
+                getHeadlineArticlesUserCase.invoke(sources, SIZE_PER_PAGE, currentPage++)
+            } else {
+                getHeadlineArticlesUserCase.invoke(query, sources, SIZE_PER_PAGE, currentPage++)
+            }
+
+            when (result) {
+                is Result.Failure -> liveData.postError(result.cause)
+                is Result.Success -> {
+                    liveData.postSuccess(result.payload)
+                    cachedArticles.addAll(result.payload)
+                }
+            }
+        }
+
+
+    companion object {
+        const val DELAY = 350L
```

Comment:
why need delay?

Reply:
it's used to implement the debounce effect. when the user typed in a query in the search input, i don't want to directly hit the API as a typed keyword change. instead i'll wait for 350ms as the user typed in a query, only after the user stop typing for more than 350ms, i do hit the API to fetch the result